### DISCUSSION TITLE
Performance of Control List in ClogTools

### DISCUSSION THREAD

**Discussion by aykaramba** [rabbibotton/clog#133] [Q&A] [answered]
I just completed the Contacts tutorial and I am starting to get addicted to Clog.  One of the things that I wanted to do is build out a panel that would render all of the widgets and give me an opportunity to wire them up and learn how they all worked in preperation to try and build a humble ticket tracker.

As part of the process of just dropping widgets on the canvas, I did notice that overall the speed of everything was actually decent and it did not slow down too too much the more widgets that were on the canvas.   But, I did notice that I had to be careful not to get too impatient and click too often on the canvas to drop / select / reselect the widgets because the Control List window would reload  and redraw for each click.

Dr. Botton, any chance you can comment on your thoughts on the overall performance of the builder tool and if there might conceptually be a way to address the performance issue?  It's not really an issue on smaller projects, but even more humble layouts require that we be careful about the click / reclick / reselect amounts and time and wait for the Control List windows to finish reloading / redrawing in order to have a predictable experience.

I would expect this to become more of an issue as we try to build more complex layouts with the number of widgets on screen multiplying.

I have attached a screenshot from my large monitor just to give you a sense of the test, although not everything is fully hooked up yet:

![Screenshot 2022-05-16 at 00-21-34 CLOG Builder](https://user-images.githubusercontent.com/16431651/168540761-816199db-57cb-464a-b6ea-5fcb31ea6c99.png)

Note: This is not a show stopper and does not need to be considered right now.  I am just learning and would like to understand what I can expect of my user experience as I move forward with a humble ticket tracker project.

Thanks.

**Comment by rabbibotton** [comment, marked as answer]
I have been busy with the tutorials then clog-web-* stuff but I will be back on the builder soon. I have for sure plans to offer better alignment and much improved performance. Part of what I have been doing for the web stuff has been improving certain areas under the covers that are in prep for that.

**Reply by aykaramba** [reply]
Thank you kindly.

**Comment by rabbibotton** [comment]
I look forward to see your creation :) BTW when you are done with the widget sampler maybe we can include it in CLOG :)

**Reply by aykaramba** [reply]
That would be awesome.

**Comment by rabbibotton** [comment]
You should now see an astronomical speed up in the builder. Let me know what you think :)

**Reply by rabbibotton** [reply]
In particular let me know where still seeing slow downs

**Comment by aykaramba** [comment]
Hello Rabbi!

Thank you very kindly for for this update!  It is very much appreciated.  Updated test details:

TEST DEFINITION:

1) On a 4k display set browser to full desktop resolution.

2) On a 4k display set each of the following builder panels 100% vertically aligned side by side:  Control Pallete, Panel, Control List, Control Properties, Control Events.  All panels need to be 100% tall aligned next to each other in a window, but they don't have to be very wide for the test just reasonable to view the labels and widgets.

3) Put an instance of every single widget on the Panel.

4) Test 1 - In the Panel, click on as many widgets as you can as fast you can to see how fast each of the other Panels re-draw.

5) Test 2 - In the Control List, click on as many widgets as you can as fast as you can to see how fast each of the other Panels re-draw.

6) Browsers tested: Nyxt, Firefox, Brave, Chromium.

7) Results:

a) WOW!  This is a huge improvement.  Never say never, but, I don't anticipate this kind of workflow to be a regular need for developer so the speed of everything is absolutely great now.

b) The fastest browser on my system felt like Nyxt the slowest felt like  Firefox.

c) You can still get the browser to struggle to keep up with re-rendering the panels if you perform each test really fast, however, even on 4k monitors with a ton of widgets, this is completely usable.  Most of the time we won't be moving widgets around but click select and then spend most of the time in the Properties, Events and List Panels to adjust components.  So the test is not totally fair, but it did test CLOG + browser rendering speed.

VERDICT: I would consider the performance issue resolved.  This is amazing, thank you.
